src/coreclr/jit/optimizer.cpp

```diff
@@ -4575,6 +4575,272 @@ PhaseStatus Compiler::optUnrollLoops()
+            {
+                case GT_ASG:
+                {
+                    GenTree* op1 = tree->gtGetOp1();
+                    GenTree* op2 = tree->gtGetOp2();
+
+                    // Only one per assignment per block can be conditionally executed.
+                    if (asgNode != nullptr || op2->OperIs(GT_SELECT))
+                    {
+                        return false;
+                    }
+
+                    // Ensure the destination of the assign is a local variable with integer type.
+                    if (!op1->OperIs(GT_LCL_VAR) || !varTypeIsIntegralOrI(op1))
+                    {
+                        return false;
+                    }
+
+                    // Ensure the nodes of the assign won't cause any additional side effects.
+                    if ((op1->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0 ||
+                        (op2->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+                    {
+                        return false;
+                    }
+
+                    asgNode = tree;
+                    asgStmt = stmt;
+                    break;
+                }
+
+                // These do not need conditional execution.
+                case GT_NOP:
+                    if (tree->gtGetOp1() != nullptr || (tree->gtFlags & (GTF_SIDE_EFFECT | GTF_ORDER_SIDEEFF)) != 0)
+                    {
+                        return false;
+                    }
+                    break;
+
+                // Cannot optimise this block.
+                default:
+                    return false;
+            }
+        }
+
+        middleBlock = middleNext;
+    }
+    if (asgNode == nullptr)
+    {
+        // The blocks checked didn't contain any ASG nodes.
+        return false;
+    }
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        JITDUMP("Conditionally executing " FMT_BB " from " FMT_BB "\n", middleBlock->bbNum, block->bbNum);
+        for (BasicBlock* dumpBlock = block; dumpBlock != middleBlock->bbNext; dumpBlock = dumpBlock->bbNext)
+        {
+            fgDumpBlock(dumpBlock);
+        }
+        JITDUMP("\n");
+    }
+#endif
+
+    // Duplicate the input of the assign.
+    // This will be used as the false result of the select node.
+    GenTree* destination = gtCloneExpr(asgNode->AsOp()->gtOp1);
+    destination->gtFlags &= GTF_EMPTY;
+
+    // Duplicate the condition and invert it
+    GenTree* cond = gtCloneExpr(last->gtGetOp1());
+    cond->gtFlags |= GTF_DONT_CSE;
+    cond->gtFlags ^= GTF_RELOP_JMP_USED;
+    cond->gtFlags ^= GTF_RELOP_NAN_UN;
+    cond->gtOper = GenTree::ReverseRelop(cond->gtOper);
+
+    // Create a select node.
+    GenTreeConditional* select =
+        gtNewConditionalNode(GT_SELECT, cond, asgNode->gtGetOp2(), destination, asgNode->TypeGet());
+
+    // Use the select as the source of the assignment.
+    asgNode->AsOp()->gtOp2 = select;
+    asgNode->AsOp()->gtFlags |= (select->gtFlags & GTF_ALL_EFFECT);
```

Comment:
If you move the assign statement to the end of `block` then you can change `block` to be BBJ_ALWAYS` and handle a more general flow pattern.

(in fact you're not too far from handling the case where the conditional assignment happens on the fall through path, which be even a bit more general).

Reply:
> If you move the assign statement to the end of `block` then you can change `block` to be BBJ_ALWAYS` and handle a more general flow pattern.

Ok, that works. I'm now transforming:
```
------------ BB03 [009..00D) -> BB05 (cond), preds={BB02} succs={BB04,BB05}
STMT00004
  *  JTRUE     void   $VN.Void
  \--*  GE        int    $102
     +--*  LCL_VAR   int    V02
     \--*  CNS_INT   int    7 $46

------------ BB04 [00D..010), preds={BB03} succs={BB05}
STMT00005
  *  ASG       int    $VN.Void
  +--*  LCL_VAR   int    V00 arg0
  \--*  CNS_INT   int    5 $47

```

To:
```
------------ BB03 [009..00D) -> BB05 (always), preds={BB02} succs={BB05}
STMT00004
  *  ASG       int    $VN.Void
  +--*  LCL_VAR   int    V00 arg0
  \--*  SELECT    int
     +--*  LT        int    $102
     |  +--*  LCL_VAR   int    V02
     |  \--*  CNS_INT   int    7 $46
     +--*  CNS_INT   int    5 $47
     \--*  LCL_VAR   int    V00

------------ BB04 [00D..010), preds={} succs={BB05}
STMT00005
  *  NOP       void
```

> Also we should think a bit about whether there is any SSA update needed.

...And now I'm getting SSA related errors.

How are the SSA definitions meant to work here. AIUI, once an SSA def is created it can't be changed.

At the moment I'm cloning the original `ASG` node and then using that to replace the `JTRUE`.
* Should I now be creating a new SSA def and deleting the old one?
* Or should I moving the old SSA def somehow?
* Or should I instead not clone the `ASG` node and instead reuse it?

Reply:
>...And now I'm getting SSA related errors.


I've fixed that by getting the SSA def from the assignment that was moved and updating the block number in it.